src/bgw/job_stat.c

```diff
@@ -186,30 +189,63 @@ calculate_jitter_percent()
 /* For failures we have standard exponential backoff based on consecutive failures
  * along with a ceiling at schedule_interval * MAX_INTERVALS_BACKOFF */
 static TimestampTz
-calculate_next_start_on_failure(TimestampTz last_finish, int consecutive_failures, BgwJob *job)
+calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failures, BgwJob *job)
 {
 	float8 jitter = calculate_jitter_percent();
 	/* consecutive failures includes this failure */
 	float8 multiplier = 1 << (consecutive_failures - 1);
+	TimestampTz res;
+	volatile bool res_set = false;
+	TimestampTz last_finish = finish_time;
+	MemoryContext oldctx;
+	if (!IS_VALID_TIMESTAMP(finish_time))
+	{
+		elog(LOG, "calculate_next_start_on_failure, got bad finish_time");
+		last_finish = ts_timer_get_current_timestamp();
+	}
+	oldctx = CurrentMemoryContext;
+	BeginInternalSubTransaction("next start on failure");
+	PG_TRY();
+	{
+		/* ival = retry_period * 2^(consecutive_failures - 1)  */
+		Datum ival = DirectFunctionCall2(interval_mul,
+										 IntervalPGetDatum(&job->fd.retry_period),
+										 Float8GetDatum(multiplier));
 
-	/* ival = retry_period * 2^(consecutive_failures - 1)  */
-	Datum ival = DirectFunctionCall2(interval_mul,
-									 IntervalPGetDatum(&job->fd.retry_period),
-									 Float8GetDatum(multiplier));
-
-	/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
-	Datum ival_max = DirectFunctionCall2(interval_mul,
-										 IntervalPGetDatum(&job->fd.schedule_interval),
-										 Float8GetDatum(MAX_INTERVALS_BACKOFF));
+		/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
+		Datum ival_max = DirectFunctionCall2(interval_mul,
+											 IntervalPGetDatum(&job->fd.schedule_interval),
+											 Float8GetDatum(MAX_INTERVALS_BACKOFF));
 
-	if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
-		ival = ival_max;
+		if (DatumGetInt32(DirectFunctionCall2(interval_cmp, ival, ival_max)) > 0)
+			ival = ival_max;
 
-	/* Add some random jitter to prevent stampeding-herds, interval will be within about +-13% */
-	ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(1.0 + jitter));
+		/* Add some random jitter to prevent stampeding-herds, interval will be within about +-13%
+		 */
+		ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(1.0 + jitter));
 
-	return DatumGetTimestampTz(
-		DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+		res = DatumGetTimestampTz(
+			DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
+		res_set = true;
+		ReleaseCurrentSubTransaction();
+	}
+	PG_CATCH();
+	{
+		RollbackAndReleaseCurrentSubTransaction();
+		FlushErrorState();
```

Comment:
Could you add a  log message printout here? It will help us debug the code until it is stable. Something like:

```
ErrorData  *errdata = CopyErrorData();
ereport(LOG_SERVER_ONLY,
        (errcode(ERRCODE_INTERNAL_ERROR),
		 errmsg_internal("calculating next start on failure did not complete"),
         errdetail("Error: %s", errdata->message));

Reply:
The server error message just says timestamp out of range. So not very helpful. I am going to make additional changes to this function in the follow up PR. Will address error logging there.